core/arch/arm/kernel/spmc_sp_handler.c

```diff
@@ -0,0 +1,195 @@
+		dst = NULL;
+		ffa_set_error(args, FFA_INVALID_PARAMETERS);
+		return NULL;
+	}
+
+	res = ffa_get_dst(args, caller_sp, &dst);
+	if (res) {
+		/*Tried to send message to an incorrect endpoint*/
+		ffa_set_error(args, res);
+		return caller_sp;
+	}
+	if (!dst) {
+		EMSG("Request to normal world not supported!");
+		ffa_set_error(args, FFA_NOT_SUPPORTED);
+		return NULL;
+	}
+
+	cpu_spin_lock(&dst->spinlock);
+	if (dst->state != sp_idle) {
+		EMSG("SP is busy");
+		ffa_set_error(args, FFA_BUSY);
+		cpu_spin_unlock(&dst->spinlock);
+		return caller_sp;
+	}
+
+	dst->state = sp_busy;
+	cpu_spin_unlock(&dst->spinlock);
+
+	/*
+	 * Store the calling endpoint id. This will make it possible to check
+	 * if the response is send back to the correct endpoint.
+	 */
+	dst->caller_id = FFA_SRC(args->a1);
+
+	/* Forward the message to the destination Secure Partition. */
+	res = sp_enter(args, dst);
+	if (res) {
+		/* The SP Panicked */
+		ffa_set_error(args, FFA_ABORTED);
+		/* Return error to calling SP*/
+		return caller_sp;
+	}
+	return dst;
+}
+
+static struct sp_session
+*ffa_handle_sp_direct_resp(struct thread_smc_args *args,
+			   struct sp_session *caller_sp)
+{
+	struct sp_session *dst = NULL;
+	TEE_Result res = FFA_OK;
+
+	if (!caller_sp) {
+		EMSG("Response from normal world not supported!");
+		ffa_set_error(args, FFA_NOT_SUPPORTED);
+		return NULL;
+	}
+
+	res = ffa_get_dst(args, caller_sp, &dst);
+	if (res) {
+		/*Tried to send response to an incorrect endpoint*/
+		ffa_set_error(args, res);
+		return caller_sp;
+	}
+
+	if (caller_sp->state != sp_busy) {
+		EMSG("SP is not waiting for a request!");
+		ffa_set_error(args, FFA_INVALID_PARAMETERS);
+		return caller_sp;
+	}
+
+	if (caller_sp->caller_id != FFA_DST(args->a1)) {
+		EMSG("FFA_MSG_SEND_DIRECT_RESP to incorrect SP!");
+		ffa_set_error(args, FFA_INVALID_PARAMETERS);
+		return caller_sp;
+	}
+
+	caller_sp->caller_id = 0;
+
+	cpu_spin_lock(&caller_sp->spinlock);
+	caller_sp->state = sp_idle;
+	cpu_spin_unlock(&caller_sp->spinlock);
+
+	if (!dst) {
+		/* Send message back to the NW*/
+		return dst;
+	}
+
+	/* Forward the message to the destination Secure Partition. */
+	res = sp_enter(args, dst);
+	if (res) {
+		/* The SP Panicked */
+		ffa_set_error(args, FFA_ABORTED);
+		/* Return error to calling SP*/
+		return caller_sp;
+	}
+	return dst;
+}
+
+/*
+ * FF-A messages handler for SP. Every messages for or by a SP is handled here.
+ * This is the entry of the sp_spmc kernel thread. The caller_sp is set to NULL
+ * when it is the Normal World.
+ */
+void spmc_sp_msg_handler(struct thread_smc_args *args,
+			 struct sp_session *caller_sp)
+{
+	thread_check_canaries();
+	do {
+		switch (args->a0) {
+		case FFA_MSG_SEND_DIRECT_REQ_32:
+			caller_sp = ffa_handle_sp_direct_req(args, caller_sp);
+			break;
+		case FFA_MSG_SEND_DIRECT_RESP_32:
+			caller_sp = ffa_handle_sp_direct_resp(args, caller_sp);
+			break;
+		case FFA_ERROR:
+			if (sp_enter(args, caller_sp))
+				panic();
```

Comment:
Done

Reply:
I really wonder if a failing SP should really mandate OP-TEE core to crash. There may be other OP-TEE services fully reliable that could still be used. This may depend on the SP and the service it provides to  secure world.
If the goal is to restart the SP, is it  really needed to reboot also OP-TEE secure world?

Reply:
If we can't send the error to the SP it's clearly broken and should be marked as dead. Then we can unwind the call chain by one link and send the error to the one calling the broken SP instead.